File: Android/app/src/main/java/com/dkaluta/prosary/models/PrayerKind.kt
package com.dkaluta.prosary.models

import androidx.annotation.StringRes
import com.dkaluta.prosary.R

/** Discriminant for the type of a saved prayer session. Only the Rosary (deeply configurable,
 * options/calendar-driven) and the Jesus Prayer (a repetition counter with no steps) warrant
 * their own cases; every other devotion is [Custom]. */
enum class PrayerKind {
    Rosary,
    JesusPrayer,

    /** Any devotion whose entire step sequence comes from a bundle's `devotion.json` instead of
     * a hardcoded engine builder — see
     * [com.dkaluta.prosary.engine.PrayerEngine.buildCustomDevotionSteps] and
     * [Prayer.customDevotionId]. One case covers every such devotion (Angelus, Stations of the
     * Cross, Franciscan Crown, Seven Sorrows, Divine Mercy Chaplet, Trisagion, ...); adding
     * another doesn't need a new [PrayerKind] case, only a new bundle. [displayName]/
     * [defaultName] below return a generic fallback for this case — real call sites (Home,
     * Favorites) read the actual devotion's name/icon from
     * [com.dkaluta.prosary.content.prayerpack.PrayerPackStore.info] instead, since a single
     * [PrayerKind] value can't carry per-bundle data. */
    Custom;

    @get:StringRes
    val displayNameRes: Int
        get() = when (this) {
            Rosary -> R.string.kind_rosary
            JesusPrayer -> R.string.kind_jesus_prayer
            Custom -> R.string.kind_devotion
        }

    /** The built-in kinds' names in each prayer language — the same map a bundle carries in
     * its manifest's displayNameByLanguage, kept here because the Rosary and the Jesus Prayer
     * have no manifest to carry it. Resolution mirrors localizedDisplayName exactly: the
     * prayer language (exact code, rites included, then its base), then the UI string. */
    private val namesByPrayerLanguage: Map<String, String>
        get() = when (this) {
            Rosary -> mapOf("he" to "מחרוזת")
            JesusPrayer -> mapOf("he" to "תפילת ישוע")
            Custom -> emptyMap()
        }

    fun displayName(context: android.content.Context): String = context.getString(displayNameRes)

    fun cardTitle(context: android.content.Context, prayerLanguage: String = LanguageCatalog.resolve(null).code): PrayerCardTitle {
        val prayerCode = LanguageCatalog.uiLanguageCode(prayerLanguage)
        val base = LanguageCatalog.baseLanguage(prayerCode) ?: prayerCode
        val prayerName = if (this == Rosary) {
            com.dkaluta.prosary.content.prayerpack.PrayerPackStore.info("rosary")?.displayNameIn(prayerCode)
        } else null
        val translatedContext = com.dkaluta.prosary.content.today.TodayTranslationLanguage.localizedContext(context, base)
        return PrayerCardTitle.resolve(
            displayName(context),
            prayerName ?: namesByPrayerLanguage[prayerCode] ?: namesByPrayerLanguage[base]
                ?: translatedContext.getString(displayNameRes),
        )
    }

    /** Default name suggested when the user creates a new favorite of this kind. */
    @get:StringRes
    val defaultNameRes: Int
        get() = when (this) {
            Rosary -> R.string.kind_default_name_rosary
            JesusPrayer -> R.string.kind_jesus_prayer
            Custom -> R.string.kind_devotion
        }
}

File: iOS/Prosary/Models/PrayerKind.swift

//
//  PrayerKind.swift
//  Prosary
//
//  Discriminant for the type of a saved prayer session. Only structurally unique sessions get a
//  case: the Rosary (deeply configurable, calendar-driven), the Jesus Prayer (a repetition
//  counter, no steps), and `custom` — the single case covering every bundle-driven devotion
//  (Angelus, Stations of the Cross, Franciscan Crown, Seven Sorrows, Divine Mercy Chaplet,
//  Trisagion, and any future .prosaryprayer with a devotion.json). Adding a devotion means
//  authoring a bundle, never a new case here.
//

import Foundation

enum PrayerKind: String, CaseIterable, Codable, Hashable {
  case rosary
  case jesusPrayer

  /// Any devotion whose entire step sequence comes from a bundle's `devotion.json` — see
  /// `PrayerEngine.buildCustomDevotionSteps` and `Prayer.customDevotionId`. `displayName`/
  /// `systemImage`/`defaultName` below return a generic fallback for this case — real call sites
  /// (Home, Favorites) read the actual devotion's name/icon from `PrayerPackStore.info(for:)`,
  /// since a single `PrayerKind` value can't carry per-bundle data.
  case custom

  /// The Jesus Prayer has no pack manifest; its other interface translations are provided
  /// by the string catalog. The Rosary obtains its prayer names from its bundled manifest.
  private var namesByPrayerLanguage: [String: String] {
    switch self {
    case .rosary:      return ["he": "מחרוזת"]
    case .jesusPrayer: return ["he": "תפילת ישוע"]
    case .custom:      return [:]
    }
  }

  var displayName: String {
    namePresentation().title
  }

  func namePresentation(prayerCode: String = LanguageCatalog.resolve(nil).code,
                        showPrayerLanguage: Bool = UserDefaults.standard.bool(forKey: PrayerNamePresentation.defaultsKey)) -> PrayerNamePresentation {
    let prayerName: String
    if self == .rosary, let info = PrayerPackStore.info(for: "rosary") {
      return info.namePresentation(prayerCode: prayerCode, showPrayerLanguage: showPrayerLanguage)
    }
    prayerName = namesByPrayerLanguage[prayerCode]
      ?? LanguageCatalog.baseLanguage(of: prayerCode).flatMap { namesByPrayerLanguage[$0] }
      ?? UILanguage.text("prayerKind.\(rawValue)", language: prayerCode, fallback: interfaceDisplayName)
    return PrayerNamePresentation(interfaceTitle: interfaceDisplayName, prayerTitle: prayerName,
                                  showPrayerLanguage: showPrayerLanguage)
  }

  private var interfaceDisplayName: String {
    switch self {
    case .rosary:      return String(localized: "prayerKind.rosary", defaultValue: "Rosary")
    case .jesusPrayer: return String(localized: "prayerKind.jesusPrayer", defaultValue: "Jesus Prayer")
    case .custom:      return String(localized: "prayerKind.custom", defaultValue: "Devotion")
    }
  }

  /// Default name suggested when the user creates a new favorite of this kind.
  var defaultName: String {
    switch self {
    case .rosary:      return String(localized: "prayerKind.defaultName.rosary", defaultValue: "My Rosary")
    case .jesusPrayer: return String(localized: "prayerKind.jesusPrayer", defaultValue: "Jesus Prayer")
    case .custom:      return String(localized: "prayerKind.custom", defaultValue: "Devotion")
    }
  }

  var systemImage: String {
    switch self {
    case .rosary:      return "circle.grid.cross"
    case .jesusPrayer: return "heart"
    case .custom:      return "sparkles"
    }
  }
}
